File: apps/android/shared/src/main/java/dev/agentdeck/shared/BridgeModels.kt
package dev.agentdeck.shared

import kotlinx.serialization.Serializable

@Serializable
data class AgentChanges(val changes: List<AgentEvent> = emptyList())

@Serializable
data class AgentHistory(val events: List<AgentEvent> = emptyList())

/**
 * An incremental stream update: the agents whose rendered state changed, plus any that are gone.
 * Everything absent from it is unchanged and must be carried over from the previous snapshot.
 */
@Serializable
data class BridgeSnapshotPatch(
    val sequence: Long = 0,
    val bridge: BridgeInfo,
    val summary: Summary,
    val agents: List<Agent> = emptyList(),
    val removed: List<String> = emptyList(),
)

/** Applies a patch to the snapshot it was computed against, preserving agent order where possible. */
fun BridgeSnapshot.applyPatch(patch: BridgeSnapshotPatch): BridgeSnapshot {
    val changed = patch.agents.associateBy { it.id }
    val kept = agents.filterNot { it.id in patch.removed }.map { changed[it.id] ?: it }
    val added = patch.agents.filter { incoming -> agents.none { it.id == incoming.id } }
    return BridgeSnapshot(patch.sequence, patch.bridge, patch.summary, kept + added)
}

@Serializable
data class SlashCommand(val name: String, val description: String? = null, val source: String = "user")

@Serializable
data class SlashCommandCatalog(val commands: List<SlashCommand> = emptyList())

@Serializable
data class BridgeSnapshot(
    val sequence: Long = 0,
    val bridge: BridgeInfo,
    val summary: Summary,
    val agents: List<Agent>,
)

@Serializable
data class BridgeInfo(val status: String, val name: String, val timestamp: String)

@Serializable
data class Summary(
    val active: Int,
    val waiting: Int,
    val errors: Int,
    val tokens: Long,
    val costUsd: Double,
)

@Serializable
data class Agent(
    val id: String,
    val name: String,
    val project: String,
    /** The directory the session works in, on the bridge's machine. */
    val cwd: String? = null,
    val model: String,
    val state: String,
    val task: String,
    val objective: String? = null,
    val progress: Double? = null,
    val tokens: Long,
    val processedTokens: Long? = null,
    val costUsd: Double,
    val lastSeenAt: String,
    /** The last moment a person looked at this session on any surface, by the bridge's account. */
    val viewedAt: String? = null,
    /** The adapter's own word for its runtime — "claude", "codex", "opencode", "pi". */
    val runtime: String? = null,
    val events: List<AgentEvent> = emptyList(),
    val capabilities: List<String>? = null,
    val rateLimits: List<RateLimitWindow> = emptyList(),
    val pendingApproval: PendingApproval? = null,
    val pendingQuestion: PendingQuestion? = null,
)

@Serializable
data class AgentEvent(
    val id: String,
    val kind: String,
    val summary: String,
    val detail: String? = null,
    val createdAt: String,
    val tool: String? = null,
    val path: String? = null,
    val command: String? = null,
    val diff: String? = null,
    val options: List<String> = emptyList(),
    /**
     * Which subagent produced this, where a subagent did.
     *
     * Absent on the parent's own work, which is the whole distinction: a
     * session running three subagents used to report their tool calls mixed
     * into its own with nothing to tell them apart.
     */
    val subagentId: String? = null,
    val subagentType: String? = null,
    /** What the run was asked to do - the Task call's own wording. */
    val subagentName: String? = null,
    /** Which exchange this belongs to - the deck's thread unit. */
    val turnId: String? = null,
)

@Serializable
data class PendingApproval(val id: String, val tool: String, val detail: String, val createdAt: String, val expiresAt: String)

@Serializable
data class PendingQuestion(
    val id: String,
    val question: String,
    val options: List<String> = emptyList(),
    val createdAt: String,
    val expiresAt: String,
)

@Serializable
data class RateLimitWindow(val id: String, val label: String, val usedPercent: Double, val resetsAt: String? = null, val account: String? = null, val runtime: String? = null)

@Serializable
data class AnalyticsSnapshot(
    val range: String,
    val project: String? = null,
    val timeZone: String = "UTC",
    val generatedAt: String,
    val summary: AnalyticsSummary,
    val series: List<AnalyticsPoint>,
    val heatmap: List<ActivityDay>,
    val projects: List<ProjectUsage>,
    val runtimes: List<RuntimeUsage>,
    val limits: List<RateLimitWindow> = emptyList(),
    val filters: AnalyticsFilters,
)

@Serializable
data class TokenFacets(val uncachedInput: Long = 0, val cachedInput: Long = 0, val cacheCreation: Long = 0, val output: Long = 0, val reasoning: Long = 0)

@Serializable
data class AnalyticsSummary(val tokens: Long, val costUsd: Double, val events: Int, val sessions: Int, val activeDays: Int, val unpricedTokens: Long = 0, val costCoveragePercent: Double = 100.0, val tokenFacets: TokenFacets = TokenFacets())

@Serializable
data class AnalyticsPoint(val bucket: String, val tokens: Long, val costUsd: Double, val events: Int)

@Serializable
data class ActivityDay(val date: String, val count: Int, val tokens: Long, val costUsd: Double)

@Serializable
data class ProjectUsage(val project: String, val tokens: Long, val costUsd: Double, val events: Int, val sessions: Int)

@Serializable
data class RuntimeUsage(val runtime: String, val tokens: Long, val costUsd: Double, val events: Int)

@Serializable
data class AnalyticsFilters(val projects: List<String>)

sealed interface AnalyticsState {
    data object Loading : AnalyticsState
    data class Ready(val data: AnalyticsSnapshot) : AnalyticsState
    data class Failed(val message: String, val previous: AnalyticsSnapshot? = null) : AnalyticsState
}

@Serializable
data class ControlRequest(
    val action: String,
    val value: String? = null,
    val commandId: String? = null,
    /** Overrides the bridge's refusal to message a session blocked on an approval or question. */
    val force: Boolean? = null,
)

/** The body of a bridge refusal, such as the 409 for messaging a blocked session. */
@Serializable
data class ControlRefusal(val error: String? = null, val detail: String? = null)

@Serializable
data class ManagedResolutionRequest(val status: String, val value: Map<String, String>)

@Serializable
data class CommandReceipt(val commandId: String, val status: String, val error: String? = null, val resultSequence: Long? = null, val updatedAt: String)

@Serializable
data class PairRequest(val code: String, val deviceName: String)

@Serializable
data class PairedDevice(val id: String, val token: String, val name: String, val createdAt: String)

data class ConnectionStatus(val phase: String, val attempt: Int = 0, val retryAt: Long? = null, val error: String? = null)

sealed interface BridgeState {
    data object Loading : BridgeState
    data class Ready(val snapshot: BridgeSnapshot, val refreshedAt: Long = System.currentTimeMillis()) : BridgeState
    data class Failed(val message: String, val previous: BridgeSnapshot? = null) : BridgeState
}

/** One runtime the bridge can host and run itself, rather than only observe. */
@Serializable
data class ManagedRuntime(
    val runtime: String,
    val capabilities: List<String> = emptyList(),
    val managed: Boolean = true,
)

@Serializable
data class ManagedRuntimes(val runtimes: List<ManagedRuntime> = emptyList())

/**
 * Body for starting a bridge-hosted session. The `cwd` must be absolute and
 * exist on the bridge's machine, so a surface offers only paths it already saw
 * a session run in - the ones the bridge has proven it can reach.
 */
@Serializable
data class ManagedSessionRequest(
    val project: String,
    val cwd: String,
    val model: String? = null,
    val objective: String? = null,
    val prompt: String? = null,
    val permissionMode: String? = null,
)

/** What a caller gets back once a hosted session is running. */
@Serializable
data class StartedManagedSession(
    val agentId: String,
    val providerSessionId: String? = null,
    val project: String,
    val model: String,
    val permissionMode: String,
)

File: apps/ios/AgentDeck/Model/BridgeModels.swift
import Foundation

/// The wire contract, mirrored from `apps/android/shared/.../BridgeModels.kt`.
///
/// Every optional here is optional on the wire too: the adapters are separate
/// programs and the bridge accepts a field absent or null. Decoding is
/// deliberately forgiving about what it does not know — an unknown key is not
/// an error, because a bridge is allowed to grow one before this app does.
/// Absent is not the same as wrong.
///
/// Swift's synthesized `init(from:)` does not fall back to a property's default
/// value when a key is missing — it throws. The bridge legitimately omits
/// `options`, `capabilities`, `events` and more, so every optional-on-the-wire
/// field is read through this rather than declared and hoped for.
private extension KeyedDecodingContainer {
    func value<T: Decodable>(_ key: Key, or fallback: T) -> T {
        ((try? decodeIfPresent(T.self, forKey: key)) ?? nil) ?? fallback
    }

    func optional<T: Decodable>(_ key: Key) -> T? {
        (try? decodeIfPresent(T.self, forKey: key)) ?? nil
    }
}

struct BridgeSnapshot: Decodable, Equatable {
    var sequence: Int64
    var bridge: BridgeInfo
    var summary: Summary
    var agents: [Agent]

    init(sequence: Int64, bridge: BridgeInfo, summary: Summary, agents: [Agent]) {
        self.sequence = sequence
        self.bridge = bridge
        self.summary = summary
        self.agents = agents
    }

    private enum CodingKeys: String, CodingKey { case sequence, bridge, summary, agents }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        sequence = container.value(.sequence, or: 0)
        bridge = try container.decode(BridgeInfo.self, forKey: .bridge)
        summary = container.value(.summary, or: Summary())
        agents = container.value(.agents, or: [])
    }
}

struct BridgeInfo: Decodable, Equatable {
    var status: String
    var name: String
    var timestamp: String

    private enum CodingKeys: String, CodingKey { case status, name, timestamp }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        status = container.value(.status, or: "unknown")
        name = container.value(.name, or: "Bridge")
        timestamp = container.value(.timestamp, or: "")
    }
}

struct Summary: Decodable, Equatable {
    var active = 0
    var waiting = 0
    var errors = 0
    var tokens: Int64 = 0
    var costUsd: Double = 0

    init() {}

    private enum CodingKeys: String, CodingKey { case active, waiting, errors, tokens, costUsd }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        active = container.value(.active, or: 0)
        waiting = container.value(.waiting, or: 0)
        errors = container.value(.errors, or: 0)
        tokens = container.value(.tokens, or: 0)
        costUsd = container.value(.costUsd, or: 0)
    }
}

struct Agent: Decodable, Equatable, Identifiable {
    var id: String
    var name: String
    var project: String
    var model: String
    var state: String
    var task: String
    /// The directory the session works in, on the bridge's machine — what lets
    /// this app offer "start another one here". Older bridges omit it.
    var cwd: String?
    var objective: String?
    var progress: Double?
    var tokens: Int64
    var processedTokens: Int64?
    var costUsd: Double
    var lastSeenAt: String
    /// The last moment a person looked at this session on any surface — the
    /// bridge's word, not this phone's. `SeenPolicy` merges it with the local
    /// marks so a session read on the watch stops badging here too.
    var viewedAt: String?
    /// The adapter's own word for its runtime — "claude", "codex", "opencode", "pi".
    var runtime: String?
    var events: [AgentEvent]
    /// An absent list is not an empty one: nil means the runtime advertised
    /// nothing, which `supportsCapability` answers false for either way, but
    /// the distinction is the wire's and is kept.
    var capabilities: [String]?
    var rateLimits: [RateLimitWindow]
    var pendingApproval: PendingApproval?
    var pendingQuestion: PendingQuestion?

    private enum CodingKeys: String, CodingKey {
        case id, name, project, model, state, task, cwd, objective, progress
        case tokens, processedTokens, costUsd, lastSeenAt, viewedAt, runtime, events
        case capabilities, rateLimits, pendingApproval, pendingQuestion
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        name = container.value(.name, or: "")
        project = container.value(.project, or: "")
        model = container.value(.model, or: "")
        state = container.value(.state, or: "idle")
        task = container.value(.task, or: "")
        cwd = container.optional(.cwd)
        objective = container.optional(.objective)
        progress = container.optional(.progress)
        tokens = container.value(.tokens, or: 0)
        processedTokens = container.optional(.processedTokens)
        costUsd = container.value(.costUsd, or: 0)
        lastSeenAt = container.value(.lastSeenAt, or: "")
        viewedAt = container.optional(.viewedAt)
        runtime = container.optional(.runtime)
        events = container.value(.events, or: [])
        capabilities = container.optional(.capabilities)
        rateLimits = container.value(.rateLimits, or: [])
        pendingApproval = container.optional(.pendingApproval)
        pendingQuestion = container.optional(.pendingQuestion)
    }
}

struct AgentEvent: Decodable, Equatable, Identifiable {
    var id: String
    var kind: String
    var summary: String
    var detail: String?
    var createdAt: String
    var tool: String?
    var path: String?
    var command: String?
    var diff: String?
    var options: [String]
    /// Which subagent produced this, where a subagent did.
    ///
    /// Claude Code tags every tool hook made inside a subagent with its own id
    /// and type. A session running three of them reported their tool calls
    /// mixed into its own, so "what is it doing" had no answer smaller than
    /// all of it.
    var subagentId: String?
    var subagentType: String?
    /// What the run was asked to do — the Task call's own wording.
    var subagentName: String?
    /// The exchange this event belongs to — one instruction and everything
    /// done in its service. The deck's thread unit; older bridges omit it.
    var turnId: String?

    private enum CodingKeys: String, CodingKey {
        case id, kind, summary, detail, createdAt, tool, path, command, diff, options
        case subagentId, subagentType, subagentName, turnId
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        kind = container.value(.kind, or: "")
        summary = container.value(.summary, or: "")
        detail = container.optional(.detail)
        createdAt = container.value(.createdAt, or: "")
        tool = container.optional(.tool)
        path = container.optional(.path)
        command = container.optional(.command)
        diff = container.optional(.diff)
        options = container.value(.options, or: [])
        subagentId = container.optional(.subagentId)
        subagentType = container.optional(.subagentType)
        subagentName = container.optional(.subagentName)
        turnId = container.optional(.turnId)
    }
}

struct PendingApproval: Decodable, Equatable {
    var id: String
    var tool: String
    var detail: String
    var createdAt: String
    var expiresAt: String

    private enum CodingKeys: String, CodingKey { case id, tool, detail, createdAt, expiresAt }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        tool = container.value(.tool, or: "")
        detail = container.value(.detail, or: "")
        createdAt = container.value(.createdAt, or: "")
        expiresAt = container.value(.expiresAt, or: "")
    }
}

struct PendingQuestion: Decodable, Equatable {
    var id: String
    var question: String
    var options: [String]
    var createdAt: String
    var expiresAt: String

    init(id: String, question: String, options: [String], createdAt: String, expiresAt: String) {
        self.id = id
        self.question = question
        self.options = options
        self.createdAt = createdAt
        self.expiresAt = expiresAt
    }

    private enum CodingKeys: String, CodingKey { case id, question, options, createdAt, expiresAt }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        question = container.value(.question, or: "")
        options = container.value(.options, or: [])
        createdAt = container.value(.createdAt, or: "")
        expiresAt = container.value(.expiresAt, or: "")
    }
}

struct RateLimitWindow: Decodable, Equatable, Identifiable {
    var id: String
    var label: String
    var usedPercent: Double
    var resetsAt: String?
    var account: String?
    var runtime: String?

    private enum CodingKeys: String, CodingKey { case id, label, usedPercent, resetsAt, account, runtime }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        label = container.value(.label, or: "")
        usedPercent = container.value(.usedPercent, or: 0)
        resetsAt = container.optional(.resetsAt)
        account = container.optional(.account)
        runtime = container.optional(.runtime)
    }
}

/// An incremental stream update: the agents whose rendered state changed, plus
/// any that are gone. Everything absent from it is unchanged and must be
/// carried over from the previous snapshot.
struct BridgeSnapshotPatch: Decodable {
    var sequence: Int64
    var bridge: BridgeInfo
    var summary: Summary
    var agents: [Agent]
    var removed: [String]

    private enum CodingKeys: String, CodingKey { case sequence, bridge, summary, agents, removed }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        sequence = container.value(.sequence, or: 0)
        bridge = try container.decode(BridgeInfo.self, forKey: .bridge)
        summary = container.value(.summary, or: Summary())
        agents = container.value(.agents, or: [])
        removed = container.value(.removed, or: [])
    }
}

extension BridgeSnapshot {
    /// Applies a patch to the snapshot it was computed against, preserving
    /// agent order where possible.
    func applying(_ patch: BridgeSnapshotPatch) -> BridgeSnapshot {
        let changed = Dictionary(patch.agents.map { ($0.id, $0) }, uniquingKeysWith: { _, last in last })
        let removed = Set(patch.removed)
        let kept = agents.filter { !removed.contains($0.id) }.map { changed[$0.id] ?? $0 }
        let known = Set(agents.map(\.id))
        let added = patch.agents.filter { !known.contains($0.id) }
        return BridgeSnapshot(
            sequence: patch.sequence,
            bridge: patch.bridge,
            summary: patch.summary,
            agents: kept + added
        )
    }
}

struct AgentHistory: Decodable {
    var events: [AgentEvent]

    private enum CodingKeys: String, CodingKey { case events }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        events = ((try? container.decodeIfPresent([AgentEvent].self, forKey: .events)) ?? nil) ?? []
    }
}

struct AgentChanges: Decodable {
    var changes: [AgentEvent]

    private enum CodingKeys: String, CodingKey { case changes }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        changes = ((try? container.decodeIfPresent([AgentEvent].self, forKey: .changes)) ?? nil) ?? []
    }
}

struct PairRequest: Encodable {
    var code: String
    var deviceName: String
}

struct PairedDevice: Decodable {
    var id: String
    var token: String
    var name: String
    var createdAt: String
}

struct ControlRequest: Encodable {
    var action: String
    var value: String?
    var commandId: String?
    /// Deliver even to a session blocked on an approval or question. Nil — the
    /// ordinary case — is omitted from the wire entirely.
    var force: Bool?
}

/// `POST /agents/:id/requests/:requestId/resolve`. A device may only ever send
/// `answered`; every other status is the runtime credential's to record, and
/// the bridge answers 403 if a device tries.
struct ResolveRequest: Encodable {
    var status: String = "answered"
    var value: [String: String]
}

/// One runtime the bridge can host and run itself, rather than only observe.
struct ManagedRuntime: Decodable, Equatable {
    var runtime: String
    var capabilities: [String]
    var managed: Bool

    private enum CodingKeys: String, CodingKey { case runtime, capabilities, managed }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        runtime = container.value(.runtime, or: "")
        capabilities = container.value(.capabilities, or: [])
        managed = container.value(.managed, or: true)
    }
}

struct ManagedRuntimes: Decodable {
    var runtimes: [ManagedRuntime] = []

    private enum CodingKeys: String, CodingKey { case runtimes }
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        runtimes = container.value(.runtimes, or: [])
    }
}

/// Body for starting a bridge-hosted session. The `cwd` must be absolute and
/// exist on the bridge's machine.
struct ManagedSessionRequest: Encodable {
    var project: String
    var cwd: String
    var model: String?
    var objective: String?
    var prompt: String?
    var permissionMode: String?
}

/// What a caller gets back once a hosted session is running.
struct StartedManagedSession: Decodable, Equatable {
    var agentId: String
    var providerSessionId: String?
    var project: String
    var model: String
    var permissionMode: String

    private enum CodingKeys: String, CodingKey { case agentId, providerSessionId, project, model, permissionMode }
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        agentId = try container.decode(String.self, forKey: .agentId)
        providerSessionId = container.optional(.providerSessionId)
        project = container.value(.project, or: "")
        model = container.value(.model, or: "")
        permissionMode = container.value(.permissionMode, or: "default")
    }
}

/// Whether a runtime advertises an action. An absent list advertises nothing.
func supportsCapability(_ capabilities: [String]?, _ action: String) -> Bool {
    capabilities?.contains(action) == true
}

/// Order-preserving unique, for the project names already on the deck offered
/// as quick fills: the first time a project appears is the order a person sees.
extension Sequence where Element: Hashable {
    var deduplicated: [Element] {
        var seen = Set<Element>()
        return filter { seen.insert($0).inserted }
    }
}

/// One instruction still waiting for its runtime — listed so it can be taken back.
struct QueuedCommand: Decodable, Identifiable, Equatable {
    var id: String
    var agentId: String
    var action: String
    var value: String?
    var createdAt: String
}

struct QueuedCommands: Decodable {
    var commands: [QueuedCommand]
}
